library/core/src/panicking.rs

```diff
@@ -29,6 +29,67 @@
 use crate::fmt;
 use crate::panic::{Location, PanicInfo};
 
+/// The entry point for panicking with a formatted message.
+///
+/// This is designed to reduce the amount of code required at the call
+/// site as much as possible (so that `panic!()` has as low an impact
+/// on (e.g.) the inlining of other functions as possible), by moving
+/// the actual formatting into this shared place.
+#[cold]
+// If panic_immediate_abort, inline the abort call,
+// otherwise avoid inlining because of it is cold path.
+#[cfg_attr(not(feature = "panic_immediate_abort"), inline(never))]
+#[cfg_attr(feature = "panic_immediate_abort", inline)]
+#[track_caller]
+#[lang = "panic_fmt"] // needed for const-evaluated panics
+#[rustc_do_not_const_check] // hooked by const-eval
+#[rustc_const_unstable(feature = "core_panic", issue = "none")]
+pub const fn panic_fmt(fmt: fmt::Arguments<'_>) -> ! {
```

Comment:
Could you move this back to where it was? The moves makes the diff harder to check.

Reply:
The current order is not entirely random. It puts every caller before every callee, so you can read from top to bottom to get the order in which functions are called.

Reply:
Oh, that explains why it is so hard to read... I find it much easier when the functions are defined before they are used. I have to scroll up and down all the time right now, since code keeps calling functions that are only defined (and documented) much later.

Also, `const_panic_fmt` is after `panic_str` despite calling that, so the order is not entirely the way you say it is.

Reply:
All right, the reordering is a separate commit now.

I still feel like defining the two functions doing the main work first makes it a lot easier to read the rest of the file -- they establish the "base vocabulary" that everything else is using. Right now I find myself having to speculate how it all holds together until I finally reach `panic_fmt`, *then* I see what really happens, and then I have to go back to the top and re-read everything.

Basically: terms should be defined before they are used. This is obviously true when doing technical writing (I cannot first use a technical term and then 2 pages later define it for the first time). And IMO it is just as true in source code, where "terms" includes things like functions and types that the code uses to express what it does.